src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,418 @@
+additive inverse. Nevertheless, overloading 'fromInteger', '+' and '*'
+is very convenient when working with algebraic graphs; we hope that
+in future Haskell's Prelude will provide a more fine-grained class
+hierarchy for algebraic structures, which we would be able to utilise
+without violating any laws.
+-}
+newtype AdjacencyMap a = AAM
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show, Eq, Ord)
+
+-- | Check if the internal graph representation is consistent,
+-- i.e. that all edges refer to existing vertices and the graph
+-- is acyclic. It should be impossible to create an inconsistent 
+-- adjacency map.
+--
+-- @
+-- consistent 'empty'           == True
+-- consistent (1 + 2)         == True
+-- consistent (1 * 2 + 2 * 3) == True
+-- @
+consistent :: Ord a => AdjacencyMap a -> Bool
+consistent (AAM m) = AM.consistent m && AM.isAcyclic m
+
+-- | Construct the /empty acyclic graph/.
+-- Complexity: /O(1)/ time and memory.
+empty :: AdjacencyMap a
+empty = coerce AM.empty
+
+-- | Construct the graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+vertex :: a -> AdjacencyMap a
+vertex = coerce AM.vertex
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is
+-- the length of the given list.
+--
+-- @
+-- vertices []        == 'empty'
+-- vertices [x]       == 'vertex' x
+-- vertices [1, 2, 3] == 1 + 2 + 3
+-- @
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices = coerce AM.vertices
+
+-- | Perform a disjoint overlay of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edgeList (overlayD empty empty)             == []
+-- edgeList (overlayD (1 * 2 + 1 * 3) (1 * 2)) == [(Left 1,Left 2) 
+--                                                ,(Left 1,Left 3)
+--                                                ,(Right 1,Right 2)]
+-- @
+overlayD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+overlayD (AAM a) (AAM b) = AAM (AM.overlay (AM.gmap Left a) (AM.gmap Right b))
+
+-- | Perform a disjoint connect of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edgeList (connectD empty empty)     == []
+-- edgeList (connectD (1 + 2) (1 + 2)) == [(Left 1,Right 1) 
+--                                        ,(Left 1,Right 2)
+--                                        ,(Left 2,Right 1)
+--                                        ,(Left 2,Right 2)]
+-- edgeList (connectD (1 * 2) (1 * 2)) == [(Left 1,Left 2) 
+--                                        ,(Left 1,Right 1)
+--                                        ,(Left 1,Right 2) 
+--                                        ,(Left 2,Right 1) 
+--                                        ,(Left 2,Right 2) 
+--                                        ,(Right 1,Right 2)]
+-- @
```

Comment:
These examples are quite complex (lots of details). Could you simplify them? Even better -- find some algebraic properties!